### ISSUE TITLE
Performance degradation when inserting long rows through HTTP

### ISSUE THREAD

**Issue by SaltTan** [ClickHouse/ClickHouse#8441]
I noticed performance degradation for inserts from log files in 19.13.5.
Here is my test:
```
clickhouse-client -q "select arrayStringConcat(arrayMap(x->toString (cityHash64(x)) , range(1000)),' ') from numbers(100000);" >test.json
clickhouse-client -q "CREATE TABLE log (row String) ENGINE = Memory"

time curl "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.json

19.13.4.32
real	0m2.717s user	0m0.316s sys	0m1.156s
real	0m2.727s user	0m0.212s sys	0m1.264s

19.13.5.44
real	0m3.726s user	0m0.304s sys	0m1.188s
real	0m3.714s user	0m0.324s sys	0m1.152s

time clickhouse-client -q 'insert into log format CSV' <test.json

19.13.4.32
real	0m2.394s user	0m0.768s sys	0m0.668s
real	0m2.304s user	0m0.704s sys	0m0.696s

19.13.5.44
real	0m2.328s user	0m0.704s sys	0m0.684s
real	0m2.289s user	0m0.676s sys	0m0.704s
```

**Comment by alexey-milovidov** [comment]
Is it attributed to https://github.com/ClickHouse/ClickHouse/pull/4454 ?

**Comment by den-crane** [comment]
```
clickhouse-client -q "CREATE TABLE log (row String) ENGINE = Null"

19.13.4.32
for i in `seq 1 50`; do curl -s -w 'Total: %{time_total}s\n'   "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20TSV" --data-binary @test.json ; done;
Total: 1.487038s
Total: 1.462736s
Total: 1.482538s
Total: 1.465072s
Total: 1.457921s
Total: 1.450939s
Total: 1.464301s
Total: 1.461511s
Total: 1.473052s
Total: 1.491245s
Total: 1.490051s
Total: 1.474327s
Total: 1.465696s
Total: 1.502737s
Total: 1.492240s
Total: 1.507607s
Total: 1.476913s
Total: 1.476432s
Total: 1.466434s
Total: 1.469403s
Total: 1.517988s

19.13.5.44
for i in `seq 1 50`; do curl -s -w 'Total: %{time_total}s\n'   "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20TSV" --data-binary @test.json ; done;
Total: 2.105578s
Total: 2.122311s
Total: 2.140414s
Total: 2.114425s
Total: 2.113737s
Total: 2.103004s
Total: 2.117699s
Total: 2.128135s
Total: 2.132715s
Total: 2.125600s
```

**Comment by alexey-milovidov** [comment]
@den-crane Could you please also provide `perf report` comparisons?

**Comment by den-crane** [comment]
not sure how to do it.

perf record -F 99 -p $(pidof clickhouse-server) -g -- sleep 30

[perf_data.zip](https://github.com/ClickHouse/ClickHouse/files/4006242/perf_data.zip)

perf diff perf.data.19.13.4.32 perf.data.19.13.5.44

```
[bridge] with build id dccadf6fa578c2aae846a7cedfc95c00ba258abb not found, continuing without symbols
# Event 'cycles'
#
# Baseline  Delta Abs  Shared Object       Symbol
# ........  .........  ..................  .........................................................................................................................................................
#
    16.37%     -5.33%  [kernel.kallsyms]   [k] clear_page_rep
    16.82%     -4.07%  clickhouse          [.] memcpy
    10.59%     -4.05%  [kernel.kallsyms]   [k] copy_user_generic_string
    11.59%     -3.18%  [kernel.kallsyms]   [k] page_fault
     6.78%     -3.02%  clickhouse          [.] DB::readEscapedStringInto<DB::PODArray<unsigned char, 4096ul, AllocatorWithHint<false, AllocatorHints::DefaultHint, 67108864ul>, 15ul, 16ul> >
               +1.30%  [kernel.kallsyms]   [k] rcu_cblist_dequeue
     0.00%     +1.25%  [kernel.kallsyms]   [k] kmem_cache_alloc
     3.52%     -1.18%  [kernel.kallsyms]   [k] get_page_from_freelist
     0.79%     +1.09%  [kernel.kallsyms]   [k] entry_SYSCALL_64
     1.34%     -0.89%  [kernel.kallsyms]   [k] free_pcppages_bulk
               +0.81%  [kernel.kallsyms]   [k] memset
               +0.81%  [kernel.kallsyms]   [k] __call_rcu.constprop.71
     2.48%     -0.77%  [kernel.kallsyms]   [k] __handle_mm_fault
     0.15%     +0.77%  [kernel.kallsyms]   [k] do_syscall_64
               +0.75%  [kernel.kallsyms]   [k] _raw_spin_lock_irq
               +0.73%  [kernel.kallsyms]   [k] kmem_cache_alloc_trace
               +0.68%  [kernel.kallsyms]   [k] ep_send_events_proc
               +0.63%  [kernel.kallsyms]   [k] ep_item_poll.isra.18
               +0.63%  [kernel.kallsyms]   [k] ep_insert
     0.05%     +0.62%  [kernel.kallsyms]   [k] get_mem_cgroup_from_mm
     0.30%     +0.60%  [kernel.kallsyms]   [k] _raw_spin_lock
               +0.58%  [kernel.kallsyms]   [k] get_random_u32
     0.05%     +0.58%  [kernel.kallsyms]   [k] kfree
     0.05%     +0.58%  [kernel.kallsyms]   [k] kmem_cache_free
     0.89%     -0.58%  [kernel.kallsyms]   [k] handle_mm_fault
     0.14%     +0.56%  [kernel.kallsyms]   [k] read_tsc
     0.00%     +0.55%  clickhouse          [.] Poco::Net::SocketImpl::poll
     0.35%     +0.49%  [kernel.kallsyms]   [k] rcu_all_qs
               +0.45%  [kernel.kallsyms]   [k] task_work_add
     0.44%     -0.44%  [kernel.kallsyms]   [k] xfrm_policy_lookup_bytype
     0.20%     +0.43%  [kernel.kallsyms]   [k] __fget
     0.84%     -0.42%  [kernel.kallsyms]   [k] __mod_zone_page_state
     0.05%     +0.41%  [kernel.kallsyms]   [k] __fget_light
               +0.40%  [kernel.kallsyms]   [k] __alloc_file
               +0.40%  [kernel.kallsyms]   [k] new_slab
               +0.40%  [kernel.kallsyms]   [k] ktime_get_ts64
     0.79%     -0.38%  [kernel.kallsyms]   [k] aa_label_sk_perm
     0.00%     +0.36%  [kernel.kallsyms]   [k] memcg_kmem_get_cache
               +0.36%  [kernel.kallsyms]   [k] file_free_rcu
               +0.36%  [kernel.kallsyms]   [k] setup_object.isra.62
               +0.36%  [kernel.kallsyms]   [k] mutex_unlock
     0.05%     +0.36%  [kernel.kallsyms]   [k] memcg_kmem_put_cache
```

**Comment by alexey-milovidov** [comment]
I compare master vs 19.13.4 and clickhouse-client vs curl.

I don't see significant difference in performance between versions (only a slight difference):

master:
```
milovidov@milovidov-desktop:~/work/ClickHouse$ time curl "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv

real    0m2,472s
user    0m0,246s
sys     0m1,195s
milovidov@milovidov-desktop:~/work/ClickHouse$ time curl "http://localhost:8123/?input_format_parallel_parsing=0&query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv

real    0m2,553s
user    0m0,188s
sys     0m1,083s
milovidov@milovidov-desktop:~/work/ClickHouse$ time curl "http://localhost:8123/?input_format_parallel_parsing=0&query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv

real    0m2,540s
user    0m0,193s
sys     0m1,090s
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.397
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.412
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.377
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.393
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.374
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.376
milovidov@milovidov-desktop:~/work/ClickHouse$ 
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --input_format_parallel_parsing=0 --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.381
```

19.13.4:

```
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.464
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.455
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.574
milovidov@milovidov-desktop:~/work/ClickHouse$ clickhouse-client --time --query "INSERT INTO log FORMAT CSV" < test.csv
1.549
milovidov@milovidov-desktop:~/work/ClickHouse$ time curl "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv

real    0m2,308s
user    0m0,151s
sys     0m1,152s
milovidov@milovidov-desktop:~/work/ClickHouse$ time curl "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv

real    0m2,349s
user    0m0,162s
sys     0m1,161s
milovidov@milovidov-desktop:~/work/ClickHouse$ time curl "http://localhost:8123/?query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv

real    0m2,306s
user    0m0,206s
sys     0m1,079s
```

But I see significant difference between clickhouse-client and curl - need to investigate.

BTW, master is build with clang on local machine (expected to be ~5% slower than official build).

**Comment by den-crane** [comment]
seems it fixed in 20.4.3.16 but clickhouse-client became faster 

```
20.3.9.70
for i in `seq 1 50`; do curl -s -w 'Total: %{time_total}s\n'   'http://localhost:8123/?input_format_parallel_parsing=0&query=INSERT%20INTO%20log%20format%20TSV' --data-binary @test.json ; done;
Total: 2.244249s
Total: 2.251710s
Total: 2.213483s
Total: 2.279380s
Total: 2.233429s
Total: 2.214297s
Total: 2.228682s

20.3.10.75
Total: 2.244871s
Total: 2.279897s
Total: 2.360135s
Total: 2.246353s
Total: 2.244625s
Total: 2.294780s
Total: 2.270477s

20.4.3.16
Total: 1.702341s
Total: 1.677854s
Total: 1.752768s
Total: 1.687315s
Total: 1.778429s
Total: 1.771065s


time clickhouse-client --input_format_parallel_parsing=0 -t -q 'insert into log format CSV' <test.json
1.348

real	0m1.365s
user	0m0.485s
sys	0m0.624s

```

**Comment by alexey-milovidov** [comment]
After #19516 the issue is fixed completely:

Master:

```
$ for i in {1..100}; do time curl "http://localhost:8123/?input_format_parallel_parsing=0&query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv; done

real    0m1,621s
user    0m0,169s
sys     0m1,079s

real    0m1,590s
user    0m0,154s
sys     0m1,072s
^C
real    0m0,580s
user    0m0,004s
sys     0m0,574s
```

21.2 and earlier:
```
$ for i in {1..100}; do time curl "http://localhost:8123/?input_format_parallel_parsing=0&query=INSERT%20INTO%20log%20format%20CSV" --data-binary @test.csv; done

real    0m2,547s
user    0m0,184s
sys     0m1,088s

real    0m2,518s
user    0m0,168s
sys     0m1,090s
^C
real    0m2,514s
user    0m0,175s
sys     0m1,089s
```
